Step3 구현

bohyeon-n opened the pull request:
- 타입별로 몬스터 클래스를 구현함
- 몬스터 타입을 정의하기 위해 enum 클래스를 사용함 
- 게임 승자를 출력하는 기능을 추가함 
- 레이스 결과 형식을 '이름[타입] : 결과' 로 수정함
- 사용자에게 몬스터 이름과 타입을 받는 기능을 추가함 

step2 리뷰를 참고하여 수정하고 step3 요구사항을 추가하였습니다.
객체지향을 지향하고자 하였으나 객체지향이 뭔지 아직 잘 모르겠습니다...! !

bohyeon-n committed:
몬스터의 타입에 따라 이동을 다르게 구현하기 위해서 각 클래스를 생성함

각 타입의 몬스터 클래스는 Monster를 속성받아 forwardNumber()메소드를 각각 구현함

bohyeon-n committed:
몬스터 정보를 받아 몬스터를 생성하는 기능 추가

inputView에 몬스터 정보를 받는 메소드 추가함
사용자 입력에 따라 다른 타입의 몬스터를 생성함
레이스 중에 몬스터를 생성하지 않고 게임 시작전에 몬스터를 생성하도록 수정함

bohyeon-n committed:
게임의 승자를 출력하는 기능 추가

게임의 승자를 출력하기 위해서 승자 배열을 반환하는 함수와 출력하는 함수를 추가함

bohyeon-n committed:
레이스 결과 출력 형식 수정함

레이스 결과에 이름과 타입이 나오도록 수정함

bohyeon-n committed:
몬스터를 생성해주는 클래스 추가

몬스터 생성 기능을 분리하기 위해서 MonsterFactory 클래스를 추가함
몬스터의 타입을 명시적으로 사용하기 위해서 MonsterType을 추가함
수정한 사항을 MonsterMain에 적용함

bohyeon-n committed:
inputView, MonsterGame 리팩토링

게임하는 몬스터들의 집합을 나타내는 이름이 적절치 않아서 monsters로 수정함
람다 표현식에서 불필요한 매개변수를 제거하기 위해 참조 연산자를 씀
특정 문자를 반복하는 메서드를 간략하게 구현하기 위해서 repeat 메서드를 씀

bohyeon-n committed:
몬스터 정보 입력이 유효한지 체크하는 로직 추가

사용자가 ','를 기준으로 문자열 두 개를 입력했는지 체크하는 로직 추가함

bohyeon-n committed:
.gitignore에 target 폴더 추가함

wheejuni commented on src/main/java/EsperMonster.java:
```diff
@@ -0,0 +1,18 @@
+
+    public EsperMonster(String name) {
+        super(name);
+        this.type = "초능력";
```
`this.type` 이라면 어차피 받을 수 있는 타입이 한정적이었을 텐데, enum을 쓰는 건 어땠을까요. 
그리고 그 enum을 super constructor에 같이 넘기고요.

wheejuni commented on src/main/java/EsperMonster.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    @Override
+    public void move() {
+        move(forwardNumber());
+    }
+
+    @Override
+    public int forwardNumber() {
+        int randomNumber = randomNumber();
+        return randomNumber == 9 ? randomNumber(1, 99) : 0;
+    }
```
고민해볼 만한 포인트가 있네요. 
아마 `forwardNumber()` 를 오버라이드해야 할 이유가 있었다면, 몬스터에 타입별로 수행해야 할 이동거리 계산 로직이 달랐기 때문이겠죠.
그렇다면 해당 로직은 몬스터의 타입에 종속성이 걸리는 것으로 보이는데, 그렇다면 이 계산 로직은 몬스터의 type enum에 존재해도 되는 거였네요. 
방법은 여러가지이지만 저 같으면 그렇게 했을 것 같습니다. 
어쨌든 지금은 외부에서 호출되지 않을 메소드를 `public` 으로 선언하고 상속받으며 오버라이드하게 설계한 것이 그리 깔끔해 보이진 않습니다.

wheejuni commented on src/main/java/MonsterFactory.java:
```diff
@@ -0,0 +1,13 @@
+public class MonsterFactory {
+    public Monster createMonster(String name, MonsterType type) {
```
`MonsterType` 을  좀 더 활용해봅시다! 제 생각엔.....

wheejuni commented on src/main/java/MonsterGame.java:
```diff
@@ -1,71 +1,66 @@
-
-        for (int i = 0; i < tryNumber; i++) {
-            monster.move(forwardNumber(randomNumber()));
+    public List<Monster> winner() {
```
불분명한 메소드 이름이고, 명사형 메소드 이름은 좋아보이지 않네요.
좀 더 직관적이고 메소드의 행위를 정의할 이름을 생각해주세요.

wheejuni commented on src/main/java/MonsterGame.java:
```diff
@@ -1,71 +1,66 @@
+                winner.add(monster);
+                continue;
+            }
+            break;
```
1등이 여러명일 수 있다는게 의외로 까다로운 요구조건이네요.
구현하신 접근도 나쁘진 않지만, 아래 링크도 한 번 참고해 주세요.
https://stackoverflow.com/questions/36428439/how-do-i-collect-multiple-maximum-values-from-a-list

wheejuni commented on src/main/java/MonsterGame.java:
```diff
@@ -1,71 +1,66 @@
 
-    private int forwardNumber(int randomResult) {
-        return randomResult >= 4 ? 1 : 0;
+    public List<Monster> monsters() {
```
getter라면 `get` 접두사를 붙여줍니다.

wheejuni commented on src/main/java/MonsterGame.java:
```diff
@@ -1,71 +1,66 @@
-        }
-
-        return result.toString();
+        return String.valueOf(character).repeat(times);
```
다음 프로젝트부터 Java 8로 고정이 되는데 이 메소드는 쓸 수 없겠네요.
이처럼 새로운 피쳐를 사용할 때는 
- 꼭 Javadoc을 확인하고 
- **since** 항목을 더블 체크 합니다. 

이 코드의 경우 JDK 8이 인스톨되어 있는 사람은 아예 컴파일 에러가 날 겁니다. 
혼자 하는 프로젝트여서 상관이 없었지만, 앞으로 늘 유념해주세요.

wheejuni commented on src/main/java/MonsterMain.java:
```diff
@@ -1,18 +1,57 @@
+    }
+
+    private Monster monster() throws IOException {
+        MonsterFactory monsterFactory = new MonsterFactory();
```
팩토리는 시즈탱크, 벌쳐, 골리앗 등을 뽑는 시설이기때문에 대부분의 경우 짓는데 비용이 많이 듭니다. 
미네랄 200원 가스 50원이었나...

하여 대부분의 경우 Java 코드에서는 팩토리를 **정적으로 선언** 하거나 **싱글톤 객체** 로 설계해 사용합니다.
팩토리 클래스를 설계한 것은, 팩토리에서 쏟아져나올 객체 물량 러시를 가기 위함이지 공장 자체를 많이 만들기 위한 건 아니기 때문입니다. 

팩토리 클래스의 생성자를 호출해 로직이 돌 때마다 새 공장을 만들어내는 것은 좋아보이지 않습니다.
`new MonsterFactory()` 하지 않고도 객체를 만들 수 있는 방법을 고민해 주세요.

wheejuni commented on src/main/java/MonsterMain.java:
```diff
@@ -1,18 +1,57 @@
+        while (true) {
+            String[] monsterInfo = inputView.monsterInfo();
+            String monsterName = monsterInfo[0];
+            MonsterType monsterType = MonsterType.getByKey(monsterInfo[1]);
+            if (monsterType == null) {
+                continue;
+            }
```
`MonsterType.getByKey()` 를 재설계하여 `java.util.Optional` 을 활용할 수 있는 방안을 고민해 주세요.

wheejuni commented on src/main/java/MonsterMain.java:
```diff
@@ -1,18 +1,57 @@
+import java.util.List;
 
 public class MonsterMain {
+    InputView inputView;
```
접근제어자를 붙입시다! 접근제어자는 나의 친구

wheejuni commented on src/main/java/MonsterFactory.java:
```diff
@@ -0,0 +1,13 @@
+            case ESPER:
+                return new EsperMonster(name);
+        }
+        return null;
```
`null` 을 받는 쪽에서 받을 준비가 되어있나요?

wheejuni commented on src/main/java/Monster.java:
```diff
@@ -1,16 +1,48 @@
+
     private int position;
+    String name;
+    String type;
```
굳이 `package-private` 할 필드가 아닌 것 같은데 `protected` 를 붙여주세요.

wheejuni reviewed (changes requested):
새벽 한시에 풀 리퀘스트를 여는 열정 좋습니다. 
코드 전반에서 아쉬웠던 점 아래와 같이 정리해 봅니다.

1. getter/setter 는 단순한 코딩 컨벤션을 넘어.... Java 언어가 갖고 있는 일종의 규약이기도 해요. 
https://imasoftwareengineer.tistory.com/101 참고하고 getter/setter 네이밍 규칙을 꼭 꼭 준수해 주세요. 
(아 참, 위 링크에서 `Serializable` 관련 내용은 굳이 안 참고하셔도 됩니다. 아직은...) 

2. 명사형 메소드 이름을 지양하시고, 메소드의 역할과 행위를 명확히 드러내는 이름을 생각해 주십시오. 

3. 접근 제어자를 (명확한 이유가 없다면) 빼먹지 맙시다.

늦은 밤까지 수고 많으셨어요. 조금만 다듬어 봅시다!

bohyeon-n committed:
몬스터 타입을 enum으로 수정, 접근 제어자 추가

몬스터 타입을 enum으로 수정하고 super 컨스트럭터에 타입을 넘기도록 수정함
name , type의 접근 제어자를 추가하고 type() 메서드 추가함

bohyeon-n committed:
몬스터 클래스 getter, setter 추가

몬스터를 캡슐화하기 위해서 getter, setter 추가함
인자가 없는 생성자를 만들어 객체 생성 후 값을 설정할 수 있도록 함(코드 리뷰 반영)

bohyeon-n committed:
명사형 메서드 이름 수정

새로운 객체를 만들어서 반환하는 메서드 이름에 create를 붙임

bohyeon-n committed:
winner를 생성하는 메서드 수정

루프를 한 번만 돌도록 로직을 수정함

bohyeon-n committed:
createMonster 메서드를 static 메서드로 수정

객체를 만들지 않고 메서드를 사용할 수 있도록 static 메서드로 수정함

honux77 commented:
수고 많으셨어요. @bohyeon-n 리뷰 꼼꼼히 읽고 다음 단계에 반영할 수 있는 것들 반영해 보세요.
